dataikuapi/dss/ml.py

```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    def disable_categories(self, categories, enable_others=False):
+        accepted_categories = self._algo_settings[self.name]["values"].keys()
+        for category in categories:
+            assert is_basestring(category)
+            assert category in accepted_categories
+        self.set_values({category: {"enabled": False}
+                         for category in categories})
+        if enable_others:
+            self.set_values({category: {"enabled": True}
+                             for category in accepted_categories
+                             if category not in categories})
+
+
+class SingleValuedHyperparameterSettings(HyperparameterSettings):
+
+    def __init__(self, name, algo_settings, accepted_values=None, accepted_types=None):
+        super(SingleValuedHyperparameterSettings, self).__init__(name, algo_settings)
+        self.accepted_values = accepted_values
+        self.accepted_types = accepted_types
+
+    def __repr__(self):
+        if self.accepted_values is not None:
+            if is_basestring(self._algo_settings[self.name]):
+                return self.__class__.__name__ + "(hyperparameter=\"{}\", value=\"{}\", accepted_values={})".format(self.name,
+                                                                                                                    self._algo_settings[self.name],
+                                                                                                                    json.dumps(self.accepted_values))
+            else:
+                return self.__class__.__name__ + "(hyperparameter=\"{}\", value={}, accepted_values={})".format(self.name,
+                                                                                                                    self._algo_settings[self.name],
+                                                                                                                    str(self.accepted_values))
+        else:
+            return self.__class__.__name__ + "(hyperparameter=\"{}\", value={})".format(self.name, self._algo_settings[self.name])
+
+    __str__ = __repr__
+
+    def set_value(self, value):
+        if self.accepted_values is not None:
+            assert value in self.accepted_values, "Invalid value for hyperparameter {}. Must be in {}".format(self.name, json.dumps(self.accepted_values))
+        if self.accepted_types is not None:
+            assert any(isinstance(value, type) for type in self.accepted_types), "Invalid type for hyperparameter {}. Type must be one of {}".format(self.name, self.accepted_types)
+        self._algo_settings._set_single_valued_hyperparameter(self.name, value)
+
+
+class AlgorithmSettings(dict):
+
+    excluded_keys = {"enabled"}
+
+    def __init__(self, raw_settings, hyperparameter_search_params):
+        super(AlgorithmSettings, self).__init__(raw_settings)
+        self._hyperparameter_search_params = hyperparameter_search_params
+
+    def _repr_html_(self):
```

Comment:
This seems risky to iterate over the underlying dict, while we always manipulate fields otherwise.
Wouldn't it be cleaner to have a `register_hyperparameters` called at the end of the init of each algo  (that would also serve for `_get_all_hyperparameter_names`).

This would allow to leverage methods like `pretty_repr` directly written from `NumericalHyperparameterSettings` and `CategoricalHyperparameterSettings` rather than manually building them here.

Reply:
Thanks for this comment, the iteration is quite clumsy indeed.
How about 
```python
class AlgorithmSettings(dict):

    excluded_keys = {"enabled"}

    def __init__(self, raw_settings, hyperparameter_search_params):
        super(AlgorithmSettings, self).__init__(raw_settings)
        self._hyperparameter_search_params = hyperparameter_search_params
        self._hyperparameters_registry = dict()

    def register_numerical_hyperparameter(self, name):
        self._hyperparameters_registry[name] = NumericalHyperparameterSettings(name, self)
        return self._hyperparameters_registry[name]

    def register_categorical_hyperparameter(self, name):
        self._hyperparameters_registry[name] = CategoricalHyperparameterSettings(name, self)
        return self._hyperparameters_registry[name]

    def register_single_valued_categorical_hyperparameter(self, name, accepted_values=None):
        self._hyperparameters_registry[name] = SingleValuedHyperparameterSettings(name, self, accepted_values=accepted_values)
        return self._hyperparameters_registry[name

    def register_single_valued_hyperparameter(self, name, accepted_types=None):
        self._hyperparameters_registry[name] = SingleValuedHyperparameterSettings(name, self, accepted_types=accepted_types)
        return self._hyperparameters_registry[name]

    def _repr_html_(self):
        res = "<pre>" + self.__class__.__name__ + "\n"
        for name, hyperparam_settings in self._hyperparameters_registry.items():
            res += hyperparam_settings._pretty_repr() + "\n"
        return res + "<details><pre>{}</pre></details>".format(self.__repr__())

    def __repr__(self):
        return self.__class__.__name__ + "(values={})".format(json.dumps(super(AlgorithmSettings, self).copy(), indent=4))

    __str__ = __repr__

    def _get_all_hyperparameter_names(self):
        return list(self._hyperparameters_registry.keys())
    ...
```
and then each algo settings would become
```python
class LogitSettings(AlgorithmSettings):

    def __init__(self, raw_settings, hyperparameter_search_params):
        super(LogitSettings, self).__init__(raw_settings, hyperparameter_search_params)
        self.C = self.register_numerical_hyperparameter("C")
        self.penalty = self.register_categorical_hyperparameter("penalty")
        self.multi_class = self.register_single_valued_categorical_hyperparameter("multi_class", accepted_values=["multinomial", "ovr"])
        self.n_jobs = self.register_single_valued_hyperparameter("n_jobs", accepted_types=[int])
```